odohlistener.go

```diff
@@ -0,0 +1,250 @@
+package rdns
+
+import (
+	"bytes"
+	"crypto/rand"
+	"crypto/tls"
+	"encoding/hex"
+	"errors"
+	"io"
+	"log/slog"
+	"net/http"
+
+	"github.com/cisco/go-hpke"
+	odoh "github.com/cloudflare/odoh-go"
+	"github.com/miekg/dns"
+)
+
+const (
+	// HPKE constants
+	kemID  = hpke.DHKEM_X25519
+	kdfID  = hpke.KDF_HKDF_SHA256
+	aeadID = hpke.AEAD_AESGCM128
+
+	defaultSeedLength = 32
+)
+
+// ODoHListener is an Oblivious DNS over HTTPS listener.
+type ODoHListener struct {
+	id          string
+	addr        string
+	proxyClient *http.Client // Client for proxy forwarding if acting as ODoH proxy
+	doh         *DoHListener
+
+	r           Resolver // Forwarding DNS queries if acting as ODoH Target
+	opt         ODoHListenerOptions
+	odohKeyPair odoh.ObliviousDoHKeyPair
+}
+
+type ODoHListenerOptions struct {
+	ListenOptions
+
+	OdohMode  string
+	AllowDoH  bool
+	KeySeed   string
+	TLSConfig *tls.Config
+}
+
+var _ Listener = &ODoHListener{}
+
+// NewODoHListener returns an instance of an oblivious DNS-over-HTTPS listener.
+func NewODoHListener(id, addr string, opt ODoHListenerOptions, resolver Resolver) (*ODoHListener, error) {
+	keyPair, err := getKeyPair(opt)
+	if err != nil {
+		Log.Error("Failed to generate HPKE key pair", "error", err)
+		return nil, err
+	}
+
+	dohOpt := DoHListenerOptions{
+		TLSConfig: opt.TLSConfig,
+		isChild:   true,
+	}
+	dohListen, err := NewDoHListener(id, addr, dohOpt, resolver)
+	if err != nil {
+		Log.Error("Failed to spawn DoH listener", "error", err)
+		return nil, err
+	}
+
+	l := &ODoHListener{
+		id:          id,
+		addr:        addr,
+		r:           resolver,
+		opt:         opt,
+		proxyClient: &http.Client{},
+		odohKeyPair: keyPair,
+		doh:         dohListen,
+	}
+
+	switch opt.OdohMode {
+	case "proxy":
+		http.HandleFunc(ODOH_PROXY_PATH, l.ODoHproxyHandler)
+	case "target":
+		http.HandleFunc(ODOH_QUERY_PATH, l.ODoHqueryHandler)
+		http.HandleFunc(ODOH_CONFIG_PATH, l.configHandler)
+	default:
+		http.HandleFunc(ODOH_PROXY_PATH, l.ODoHproxyHandler)
+		http.HandleFunc(ODOH_QUERY_PATH, l.ODoHqueryHandler)
+		http.HandleFunc(ODOH_CONFIG_PATH, l.configHandler)
+	}
+	return l, nil
+}
+
+func (s *ODoHListener) Start() error {
+	return s.doh.Start()
+}
+
+func (s *ODoHListener) Stop() error {
+	return s.doh.Stop()
+}
+
+func (s *ODoHListener) ODoHproxyHandler(w http.ResponseWriter, r *http.Request) {
+	if r.Method != "POST" {
+		http.Error(w, "only POST allowed", http.StatusMethodNotAllowed)
+		return
+	}
```

Comment:
A further advantage of using a `*http.ServeMux` is that you can specify the HTTP method it should handle in the pattern, making this check unnecessary. See https://pkg.go.dev/net/http#ServeMux